examples/mediadevices/src/helpers.js

```diff
@@ -97,12 +97,16 @@ function applyVideoInputDeviceSelection(deviceId, video, room) {
  * @property {Array<MediaDeviceInfo>} videoinput
  */
 function getDeviceSelectionOptions() {
-  return navigator.mediaDevices.enumerateDevices().then(function(deviceInfos) {
-    var kinds = ['audioinput', 'audiooutput', 'videoinput'];
-    return kinds.reduce(function(deviceSelectionOptions, kind) {
-      deviceSelectionOptions[kind] = getDevicesOfKind(deviceInfos, kind);
-      return deviceSelectionOptions;
-    }, {});
+  // before calling enumerateDevices, get media permissions (.getUserMedia)
+  // w/o media permissions, browsers do not return device Ids.
+  return navigator.mediaDevices.getUserMedia({ audio: true, video: true }).then(() => {
```

Comment:
@makarandp0 I think we should do something like:
```js
/**
 * Ensure that media permissions are obtained.
 * @returns {Promise<void>}
 */
function ensureMediaPermissions() {
  return navigator.mediaDevices.enumerateDevices().then(function(devices) {
    return devices.every(function(device) {
      return device.deviceId === '';
    });
  }).then(function(shouldAskForMediaPermissions) {
    if (!shouldAskForMediaPermissions) {
      return;
    }
    return navigator.mediaDevices.getUserMedia({ audio: true, video: true }).then(function(mediaStream) {
      mediaStream.getTracks().forEach(function(track) {
        track.stop();
      });
    });
  });
}

function getDeviceSelectionOptions() {
  // NOTE: Chrome 81+, Firefox and Safari all require media
  // permissions to be enabled for enumerateDevices() to
  // return the proper device IDs.
  return ensureMediaPermissions().then(function() {
    return navigator.mediaDevices.enumerateDevices();
  }).then(function(deviceInfos) {
    ...
  });
}
```

Reply:
@manjeshbhargav - The behavior of safari v/s chrome is slightly different when called w/o media permissions. (safari  does not provide labels for devices/ chrome does not provide device ids). We would have to ensure that we check the exact expected missing properties depending on browser. 

I think simply calling `navigator.mediaDevices.getUserMedia` addresses the problem nicely. It seem to be no-op if you already have media permissions.
 
Are you worried that media stream returned by the  `getUserMedia` would have bad effect? I think since we are not keeping reference to it, it would get garbage collected and dealed appropriately by the browser.

Reply:
@makarandp0 What Chrome has implemented is spec-compliant behavior. So, Safari will have this at some point. Further, we should make sure that the camera light is turned off after we call `getUserMedia` in order to get permissions to access device ids. What you can do is:
```js
return devices.every(function(device) {
  return !(device.id && device.label);
});
```
This will cover both Chrome and Safari use-cases.

Reply:
Thanks @manjeshbhargav - good point about the camera light. It does stay on for sometime if we do not stop the tracks. I will update per your suggestion.